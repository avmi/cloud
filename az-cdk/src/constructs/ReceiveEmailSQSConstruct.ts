import { Construct } from '@aws-cdk/core';
import { SqsSubscription } from '@aws-cdk/aws-sns-subscriptions';
import { Topic } from '@aws-cdk/aws-sns';
import { Queue } from '@aws-cdk/aws-sqs';
import { email2key } from '@cpmech/basic';
import { SESDefaultRuleSetConstruct } from '../custom-resources/SESDefaultRuleSetConstruct';

export interface IReceiveEmailProps {
  emails: string[]; // emails
  topicNames?: string[]; // set SNS topic; default is equal to email2key(example@here.com) => example_here_com
}

export class ReceiveEmailSQSConstruct extends Construct {
  readonly topicArns: string[] = [];
  readonly queueUrls: string[] = [];

  constructor(scope: Construct, id: string, props: IReceiveEmailProps) {
    super(scope, id);

    for (let i = 0; i < props.emails.length; i++) {
      const title = props.topicNames ? props.topicNames[i] : email2key(props.emails[i]);
      const topic = new Topic(this, `Topic${i}`, { topicName: title });
      const queue = new Queue(this, `Queue${i}`, { queueName: title });
      topic.addSubscription(new SqsSubscription(queue));
      this.topicArns.push(topic.topicArn);
      this.queueUrls.push(
        `https://sqs.${queue.stack.region}.amazonaws.com/${queue.stack.account}/${title}`,
      );
    }

    new SESDefaultRuleSetConstruct(this, 'DefaultRuleSet', {
      emails: props.emails,
      topicArns: this.topicArns,
    });
  }
}
