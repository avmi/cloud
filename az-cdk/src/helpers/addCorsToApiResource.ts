import { IResource, MockIntegration, PassthroughBehavior } from '@aws-cdk/aws-apigateway';

export const addCorsToApiResource = (
  resource: IResource,
  methods: string[] = ['GET', 'PUT', 'POST', 'DELETE'],
) => {
  const m = methods.join(',');
  resource.addMethod(
    'OPTIONS',
    new MockIntegration({
      integrationResponses: [
        {
          statusCode: '200',
          responseParameters: {
            'method.response.header.Access-Control-Allow-Headers': `'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'`,
            'method.response.header.Access-Control-Allow-Origin': `'*'`,
            'method.response.header.Access-Control-Allow-Credentials': `'false'`,
            'method.response.header.Access-Control-Allow-Methods': `'OPTIONS,${m}'`,
          },
        },
      ],
      passthroughBehavior: PassthroughBehavior.NEVER,
      requestTemplates: {
        'application/json': '{"statusCode": 200}',
      },
    }),
    {
      methodResponses: [
        {
          statusCode: '200',
          responseParameters: {
            'method.response.header.Access-Control-Allow-Headers': true,
            'method.response.header.Access-Control-Allow-Methods': true,
            'method.response.header.Access-Control-Allow-Credentials': true,
            'method.response.header.Access-Control-Allow-Origin': true,
          },
        },
      ],
    },
  );
};
