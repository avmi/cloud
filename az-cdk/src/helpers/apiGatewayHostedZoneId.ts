/*
  The API Gateway HostedZoneId for each region has to be looked up here:
     https://docs.aws.amazon.com/general/latest/gr/rande.html#apigateway_region

  Examples:
    US East (N. Virginia)      us-east-1       execute-api.us-east-1.amazonaws.com       Z1UJRXOUMOOFQ8
    US West (N. California)    us-west-1       execute-api.us-west-1.amazonaws.com       Z2MUQ32089INYE
    US East (Ohio)             us-east-2       execute-api.us-east-2.amazonaws.com       ZOJJZC49E0EPZ
    US West (Oregon)           us-west-2       execute-api.us-west-2.amazonaws.com       Z2OJLYMUO9EFXC
    Asia Pacific (Hong Kong)   ap-east-1       execute-api.ap-east-1.amazonaws.com       Z3FD1VL90ND7K5
    Asia Pacific (Singapore)   ap-southeast-1  execute-api.ap-southeast-1.amazonaws.com  ZL327KTPIQFUL
    Asia Pacific (Sydney)      ap-southeast-2  execute-api.ap-southeast-2.amazonaws.com  Z2RPCDW04V8134
    South America (Sao Paulo)  sa-east-1       execute-api.sa-east-1.amazonaws.com       ZCMLWB8V5SYIT
*/

export const apiGatewayHostedZoneId = (region: string) => {
  switch (region.toLowerCase()) {
    case 'us-east-2':
      return 'ZOJJZC49E0EPZ';
    case 'us-east-1':
      return 'Z1UJRXOUMOOFQ8';
    case 'us-west-1':
      return 'Z2MUQ32089INYE';
    case 'us-west-2':
      return 'Z2OJLYMUO9EFXC';
    case 'ap-east-1':
      return 'Z3FD1VL90ND7K5';
    case 'ap-south-1':
      return 'Z3VO1THU9YC4UR';
    case 'ap-northeast-2':
      return 'Z20JF4UZKIW1U8';
    case 'ap-southeast-1':
      return 'ZL327KTPIQFUL';
    case 'ap-southeast-2':
      return 'Z2RPCDW04V8134';
    case 'ap-northeast-1':
      return 'Z1YSHQZHG15GKL';
    case 'ca-central-1':
      return 'Z19DQILCV0OWEC';
    case 'eu-central-1':
      return 'Z1U9ULNL0V5AJ3';
    case 'eu-west-1':
      return 'ZLY8HYME6SFDD';
    case 'eu-west-2':
      return 'ZJ5UAJN8Y3Z2Q';
    case 'eu-west-3':
      return 'Z3KY65QIEKYHQQ';
    case 'eu-north-1':
      return 'Z3UWIKFBOOGXPP';
    case 'me-south-1':
      return 'Z20ZBPC0SS8806';
    case 'sa-east-1':
      return 'ZCMLWB8V5SYIT';
  }
  throw new Error(`cannot get API Gateway HostedId for region "${region}"`);
};
