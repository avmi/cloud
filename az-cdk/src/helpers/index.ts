export * from './apiGatewayHostedZoneId';
export * from './envars2cdk';
export * from './npmCommands';
export * from './ssmSecret';
export * from './synthApp';
