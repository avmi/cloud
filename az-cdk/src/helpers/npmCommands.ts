export interface INPMCommands {
  install: string[];
  build: string[];
}

export const npmCommands = (extraBeforeTest = '', useYarn = false): INPMCommands => {
  if (useYarn) {
    return {
      install: [
        'curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -',
        'echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list',
        'sudo apt-get update && sudo apt-get install yarn',
        'yarn',
      ],
      build: ['yarn test', 'yarn build'],
    };
  }
  return {
    install: ['npm install'],
    build: extraBeforeTest
      ? [extraBeforeTest, 'npm run test', 'npm run build']
      : ['npm run test', 'npm run build'],
  };
};
