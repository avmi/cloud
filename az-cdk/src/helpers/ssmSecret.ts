import { SecretValue, CfnDynamicReference, CfnDynamicReferenceService } from '@aws-cdk/core';

// ssmSecret creates a reference to a parameter in the SSM Parameter Store (not Secrets Manager!)
export const ssmSecret = (input: { name: string; version: string }): SecretValue => {
  return SecretValue.cfnDynamicReference(
    new CfnDynamicReference(CfnDynamicReferenceService.SSM, `${input.name}:${input.version}`),
  );
};
