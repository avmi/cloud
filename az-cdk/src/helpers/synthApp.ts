import { App } from '@aws-cdk/core';

// synthApp synthetizes App and returns object
export const synthApp = (app: App): any => {
  const assembly = app.synth();
  const synthesized = assembly.stacks[0];
  return synthesized.template;
};

// synthAppString synthetizes App and returns stringifyied object
export const synthAppString = (app: App): string => {
  return JSON.stringify(synthApp(app), undefined, 2);
};
