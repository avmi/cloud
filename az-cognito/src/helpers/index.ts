export * from './flattenAttributes';
export * from './email2username';
