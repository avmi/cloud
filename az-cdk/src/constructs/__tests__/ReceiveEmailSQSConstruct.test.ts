import { App, Stack } from '@aws-cdk/core';
import { ReceiveEmailSQSConstruct } from '../ReceiveEmailSQSConstruct';
import { synthAppString } from '../../helpers/synthApp';

describe('ReceiveEmailSQSConstruct', () => {
  it('synthetizes properly', () => {
    const app = new App();
    const stack = new Stack(app, 'Stack');
    new ReceiveEmailSQSConstruct(stack, 'ReceiveEmailSQS', {
      emails: ['admin@here.com', 'tester@here.com'],
    });
    expect(synthAppString(app)).toMatchSnapshot();
  });
});
