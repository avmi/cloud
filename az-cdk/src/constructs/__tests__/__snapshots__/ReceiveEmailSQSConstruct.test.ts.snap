// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReceiveEmailSQSConstruct synthetizes properly 1`] = `
"{
  \\"Resources\\": {
    \\"ReceiveEmailSQSTopic08E880318\\": {
      \\"Type\\": \\"AWS::SNS::Topic\\",
      \\"Properties\\": {
        \\"TopicName\\": \\"admin_here_com\\"
      }
    },
    \\"ReceiveEmailSQSQueue0389ABB93\\": {
      \\"Type\\": \\"AWS::SQS::Queue\\",
      \\"Properties\\": {
        \\"QueueName\\": \\"admin_here_com\\"
      }
    },
    \\"ReceiveEmailSQSQueue0PolicyA9BF0D7F\\": {
      \\"Type\\": \\"AWS::SQS::QueuePolicy\\",
      \\"Properties\\": {
        \\"PolicyDocument\\": {
          \\"Statement\\": [
            {
              \\"Action\\": \\"sqs:SendMessage\\",
              \\"Condition\\": {
                \\"ArnEquals\\": {
                  \\"aws:SourceArn\\": {
                    \\"Ref\\": \\"ReceiveEmailSQSTopic08E880318\\"
                  }
                }
              },
              \\"Effect\\": \\"Allow\\",
              \\"Principal\\": {
                \\"Service\\": \\"sns.amazonaws.com\\"
              },
              \\"Resource\\": {
                \\"Fn::GetAtt\\": [
                  \\"ReceiveEmailSQSQueue0389ABB93\\",
                  \\"Arn\\"
                ]
              }
            }
          ],
          \\"Version\\": \\"2012-10-17\\"
        },
        \\"Queues\\": [
          {
            \\"Ref\\": \\"ReceiveEmailSQSQueue0389ABB93\\"
          }
        ]
      }
    },
    \\"ReceiveEmailSQSQueue0StackReceiveEmailSQSTopic0539F7BF50F29907E\\": {
      \\"Type\\": \\"AWS::SNS::Subscription\\",
      \\"Properties\\": {
        \\"Protocol\\": \\"sqs\\",
        \\"TopicArn\\": {
          \\"Ref\\": \\"ReceiveEmailSQSTopic08E880318\\"
        },
        \\"Endpoint\\": {
          \\"Fn::GetAtt\\": [
            \\"ReceiveEmailSQSQueue0389ABB93\\",
            \\"Arn\\"
          ]
        }
      }
    },
    \\"ReceiveEmailSQSTopic172D60683\\": {
      \\"Type\\": \\"AWS::SNS::Topic\\",
      \\"Properties\\": {
        \\"TopicName\\": \\"tester_here_com\\"
      }
    },
    \\"ReceiveEmailSQSQueue1C13C7B57\\": {
      \\"Type\\": \\"AWS::SQS::Queue\\",
      \\"Properties\\": {
        \\"QueueName\\": \\"tester_here_com\\"
      }
    },
    \\"ReceiveEmailSQSQueue1PolicyEB4798A5\\": {
      \\"Type\\": \\"AWS::SQS::QueuePolicy\\",
      \\"Properties\\": {
        \\"PolicyDocument\\": {
          \\"Statement\\": [
            {
              \\"Action\\": \\"sqs:SendMessage\\",
              \\"Condition\\": {
                \\"ArnEquals\\": {
                  \\"aws:SourceArn\\": {
                    \\"Ref\\": \\"ReceiveEmailSQSTopic172D60683\\"
                  }
                }
              },
              \\"Effect\\": \\"Allow\\",
              \\"Principal\\": {
                \\"Service\\": \\"sns.amazonaws.com\\"
              },
              \\"Resource\\": {
                \\"Fn::GetAtt\\": [
                  \\"ReceiveEmailSQSQueue1C13C7B57\\",
                  \\"Arn\\"
                ]
              }
            }
          ],
          \\"Version\\": \\"2012-10-17\\"
        },
        \\"Queues\\": [
          {
            \\"Ref\\": \\"ReceiveEmailSQSQueue1C13C7B57\\"
          }
        ]
      }
    },
    \\"ReceiveEmailSQSQueue1StackReceiveEmailSQSTopic13E4AE9FC8AC7529D\\": {
      \\"Type\\": \\"AWS::SNS::Subscription\\",
      \\"Properties\\": {
        \\"Protocol\\": \\"sqs\\",
        \\"TopicArn\\": {
          \\"Ref\\": \\"ReceiveEmailSQSTopic172D60683\\"
        },
        \\"Endpoint\\": {
          \\"Fn::GetAtt\\": [
            \\"ReceiveEmailSQSQueue1C13C7B57\\",
            \\"Arn\\"
          ]
        }
      }
    },
    \\"ReceiveEmailSQSDefaultRuleSetFunctionServiceRole7ED6552F\\": {
      \\"Type\\": \\"AWS::IAM::Role\\",
      \\"Properties\\": {
        \\"AssumeRolePolicyDocument\\": {
          \\"Statement\\": [
            {
              \\"Action\\": \\"sts:AssumeRole\\",
              \\"Effect\\": \\"Allow\\",
              \\"Principal\\": {
                \\"Service\\": \\"lambda.amazonaws.com\\"
              }
            }
          ],
          \\"Version\\": \\"2012-10-17\\"
        },
        \\"ManagedPolicyArns\\": [
          {
            \\"Fn::Join\\": [
              \\"\\",
              [
                \\"arn:\\",
                {
                  \\"Ref\\": \\"AWS::Partition\\"
                },
                \\":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\\"
              ]
            ]
          }
        ]
      }
    },
    \\"ReceiveEmailSQSDefaultRuleSetFunctionServiceRoleDefaultPolicyAC9D685E\\": {
      \\"Type\\": \\"AWS::IAM::Policy\\",
      \\"Properties\\": {
        \\"PolicyDocument\\": {
          \\"Statement\\": [
            {
              \\"Action\\": \\"ses:CreateReceiptRule\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:CreateReceiptRuleSet\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:DeleteReceiptRule\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:DeleteReceiptRuleSet\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:DescribeActiveReceiptRuleSet\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:DescribeReceiptRuleSet\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:ListReceiptRuleSets\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:SetActiveReceiptRuleSet\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            },
            {
              \\"Action\\": \\"ses:UpdateReceiptRule\\",
              \\"Effect\\": \\"Allow\\",
              \\"Resource\\": \\"*\\"
            }
          ],
          \\"Version\\": \\"2012-10-17\\"
        },
        \\"PolicyName\\": \\"ReceiveEmailSQSDefaultRuleSetFunctionServiceRoleDefaultPolicyAC9D685E\\",
        \\"Roles\\": [
          {
            \\"Ref\\": \\"ReceiveEmailSQSDefaultRuleSetFunctionServiceRole7ED6552F\\"
          }
        ]
      }
    },
    \\"ReceiveEmailSQSDefaultRuleSetFunction35E9DAB9\\": {
      \\"Type\\": \\"AWS::Lambda::Function\\",
      \\"Properties\\": {
        \\"Code\\": {
          \\"S3Bucket\\": {
            \\"Ref\\": \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3Bucket84802930\\"
          },
          \\"S3Key\\": {
            \\"Fn::Join\\": [
              \\"\\",
              [
                {
                  \\"Fn::Select\\": [
                    0,
                    {
                      \\"Fn::Split\\": [
                        \\"||\\",
                        {
                          \\"Ref\\": \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3VersionKey04C3B8A9\\"
                        }
                      ]
                    }
                  ]
                },
                {
                  \\"Fn::Select\\": [
                    1,
                    {
                      \\"Fn::Split\\": [
                        \\"||\\",
                        {
                          \\"Ref\\": \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3VersionKey04C3B8A9\\"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        \\"Handler\\": \\"index.sesDefaultRuleSet\\",
        \\"Role\\": {
          \\"Fn::GetAtt\\": [
            \\"ReceiveEmailSQSDefaultRuleSetFunctionServiceRole7ED6552F\\",
            \\"Arn\\"
          ]
        },
        \\"Runtime\\": \\"nodejs8.10\\",
        \\"Timeout\\": 60
      },
      \\"DependsOn\\": [
        \\"ReceiveEmailSQSDefaultRuleSetFunctionServiceRoleDefaultPolicyAC9D685E\\",
        \\"ReceiveEmailSQSDefaultRuleSetFunctionServiceRole7ED6552F\\"
      ]
    },
    \\"ReceiveEmailSQSDefaultRuleSet45C1617D\\": {
      \\"Type\\": \\"AWS::CloudFormation::CustomResource\\",
      \\"Properties\\": {
        \\"ServiceToken\\": {
          \\"Fn::GetAtt\\": [
            \\"ReceiveEmailSQSDefaultRuleSetFunction35E9DAB9\\",
            \\"Arn\\"
          ]
        },
        \\"Emails\\": [
          \\"admin@here.com\\",
          \\"tester@here.com\\"
        ],
        \\"TopicArns\\": [
          {
            \\"Ref\\": \\"ReceiveEmailSQSTopic08E880318\\"
          },
          {
            \\"Ref\\": \\"ReceiveEmailSQSTopic172D60683\\"
          }
        ]
      },
      \\"UpdateReplacePolicy\\": \\"Delete\\",
      \\"DeletionPolicy\\": \\"Delete\\"
    }
  },
  \\"Parameters\\": {
    \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3Bucket84802930\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"S3 bucket for asset \\\\\\"8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168\\\\\\"\\"
    },
    \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168S3VersionKey04C3B8A9\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"S3 key for asset version \\\\\\"8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168\\\\\\"\\"
    },
    \\"AssetParameters8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168ArtifactHashDB14117A\\": {
      \\"Type\\": \\"String\\",
      \\"Description\\": \\"Artifact hash for asset \\\\\\"8ee8fb524eef3fdcf0c8f7b3a096bc2b85686756e8a3c7176c033e494f985168\\\\\\"\\"
    }
  }
}"
`;
